### PR TITLE
New pattern submission: Amazon S3, SQS, Lambda, and Bedrock CDK Python

### PR DESCRIPTION
**Pull request by pdineshbalaaji**
*Issue #, if available:*

*Description of changes:*

This pattern creates an Amazon S3 bucket that sends 'OBJECT_CREATED' events to an SQS queue. These events are then processed by a Lambda function that utilizes Amazon Bedrock to extract the summary or sentiment of the input file. The summary is subsequently stored in a different S3 bucket for consumption. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by ellisms** [comment]
Hi @pdineshbalaaji I'm reviewing this pattern and I'm unable to deploy it due to the following:
```
"Invalid request provided: Queue visibility timeout: 60 seconds is less than Function timeout: 300 seconds```
Can you please update the settings?

**Comment by pdineshbalaaji** [comment]
@ellisms I have updated the visibility timeout settings and validated the changes. Can you please check if its working from your end.

File: s3-sqs-lambda-bedrock-s3-cdk-python/README.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    cdk bootstrap
+    ```
+8. From the command line, use CDK to deploy the AWS resources for the pattern. You'll be prompted to approve security related changes during the deployment.
+    ```
+    cdk deploy
+    ```
+
+## How it works
+
+* Upload a support conversation history input file to the input S3 Bucket, triggering an 'OBJECT_CREATED' event notification to an SQS Queue. Upon receipt, the SQS Queue triggers a Lambda function which will fetch the content of the file using Bucket and Object key info in the event. The Lambda constructs a prompt containing the support conversation and invokes Amazon Titan Text Express from Bedrock to generate conversation summary and sentiment in JSON format. Finally, the Lambda uploads this summary to the summary S3 Bucket.
```
Comment (ellisms) [inline s3-sqs-lambda-bedrock-s3-cdk-python/README.md:59]:
```suggestion
* Upload a support conversation history input file to the input S3 Bucket, triggering an 'OBJECT_CREATED' event notification to an SQS Queue. A Lambda function receives the message from the queue and fetches the content  using the Bucket and Object key information in the event. The function constructs a prompt containing the support conversation and invokes Amazon Titan Text Express from Bedrock to generate a conversation summary and sentiment in JSON format. Finally, the Lambda function uploads the output to the summary S3 Bucket.
```

File: s3-sqs-lambda-bedrock-s3-cdk-python/README.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+
+## Testing
+
+1. Navigate to the S3 bucket containing key 's3sqslambdabedrocks3cdkpythons-inputbucket**' along with your stack name and a hash key
```
Comment (ellisms) [inline s3-sqs-lambda-bedrock-s3-cdk-python/README.md:64]:
```suggestion
1. Navigate to the S3 bucket with a name containing 's3sqslambdabedrocks3cdkpythons-inputbucket**' 
```

File: s3-sqs-lambda-bedrock-s3-cdk-python/README.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+2. Upload input file containing conversation history to the bucket (sample input files are available under ./lambda folder)
+3. Wait for the AWS Lambda function to invoke the Amazon Bedrock model to get the sentiment and summary of the conversation. The Response will be uploaded to 's3sqslambdabedrocks3cdkpytho-summarybucket' using the same name of the input file uploaded earlier.
```
Comment (ellisms) [inline s3-sqs-lambda-bedrock-s3-cdk-python/README.md:66]:
```suggestion
3. Wait for Amazon Bedrock to produce the sentiment and summary of the conversation. The Response will be uploaded to 's3sqslambdabedrocks3cdkpytho-summarybucket' using the same name of the input file uploaded earlier.
```

File: s3-sqs-lambda-bedrock-s3-cdk-python/lambda/index.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    s3.Object(bucket_summary, key).put(Body=summary)
+    print('Summary written to s3 bucket')
+            
+    return 'Successfully processed SQS messages'
+
+
+#function to read from S3 bucket and key as input
+def read_s3_file(bucket, key):
+    s3 = boto3.client('s3')
```
Comment (ellisms) [inline s3-sqs-lambda-bedrock-s3-cdk-python/lambda/index.py:67]: You can instantiate the client at the top-level so it takes advantage of warm-starts.

File: s3-sqs-lambda-bedrock-s3-cdk-python/s3_sqs_lambda_bedrock_s3_cdk_python/s3_sqs_lambda_bedrock_s3_cdk_python_stack.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        timeout=cdk.Duration.seconds(300),
+        )
+
+        # Configure queue as event source
+        function.add_event_source(lambda_event_sources.SqsEventSource(queue, batch_size=10))    
+
+        # Grant Lambda permission to invoke bedrock-runtime
+        function.add_to_role_policy(iam.PolicyStatement(
+            actions=["bedrock:InvokeModel"],
+            resources=["*"]
```
Comment (ellisms) [inline s3-sqs-lambda-bedrock-s3-cdk-python/s3_sqs_lambda_bedrock_s3_cdk_python/s3_sqs_lambda_bedrock_s3_cdk_python_stack.py:74]: Scope the resources to the model you are using.

**Review by ellisms — CHANGES_REQUESTED** [review]

**Comment by ellisms** [comment]
Thanks - it works now. I left feedback and requested some changes.

File: s3-sqs-lambda-bedrock-s3-cdk-python/README.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
+# Conversation Summarization: Leveraging Amazon S3, SQS, Lambda, and Bedrock\
```
[the remark this answers is not in the document]
Comment (pdineshbalaaji) [reply s3-sqs-lambda-bedrock-s3-cdk-python/README.md]: change implemented

File: s3-sqs-lambda-bedrock-s3-cdk-python/README.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+   For a Windows platform, activate the virtualenv like this:
+   ```
+   .venv\Scripts\activate.bat
+5. Install the Python required dependencies:
+   ```
+   pip install -r requirements.txt
+   ```
+6. Review the CloudFormation template the cdk generates for you stack using the following AWS CDK CLI command:
+   ```
+   cdk synth
```
[the remark this answers is not in the document]
Comment (pdineshbalaaji) [reply s3-sqs-lambda-bedrock-s3-cdk-python/README.md:46]: cdk.json added to resolve this error

File: s3-sqs-lambda-bedrock-s3-cdk-python/s3_sqs_lambda_bedrock_s3_cdk_python/s3_sqs_lambda_bedrock_s3_cdk_python_stack.py (hunk 5 above)
Comment (pdineshbalaaji) [reply s3-sqs-lambda-bedrock-s3-cdk-python/s3_sqs_lambda_bedrock_s3_cdk_python/s3_sqs_lambda_bedrock_s3_cdk_python_stack.py:74]: it depends on the region in which they deploy the pattern, so added a comment below in the code which can be uncommented to add the specific resource information

**Comment by pdineshbalaaji** [comment]
@ellisms I have implemented the requested changes

**Comment by ellisms** [comment]
Thanks! Adding this to the queue for publishing.

**Pull request merged**